safehttp/plugins/cors/cors.go

```diff
@@ -0,0 +1,217 @@
+	AllowedOrigins map[string]bool
+	allowedHeaders map[string]bool
+	// ExposedHeaders determines which headers should be set in the
+	// Access-Control-Expose-Headers header. This controls which headers in the
+	// response should be accessible by JavaScript.
+	//
+	// If ExposedHeaders is nil, then the header is not set, meaning that nothing
+	// is exposed.
+	ExposedHeaders []string
+	// AllowCredentials determines if Access-Control-Allow-Credentials should be
+	// set to true, which would allow cookies to be attached to requests.
+	AllowCredentials bool
+	// MaxAge sets the Access-Control-Max-Age header.
+	// MaxAge=0 results in MaxAge: 5.
+	// This default of 5 seconds is set because it is the default used by Chromium
+	// according to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age
+	MaxAge int
+}
+
+// Default creates a CORS Interceptor with default settings.
+// Those defaults are:
+//   - No Exposed Headers
+//   - No Allowed Headers
+//   - AllowCredentials: false
+//   - MaxAge: 5 seconds
+func Default(allowedOrigins ...string) *Interceptor {
+	ao := map[string]bool{}
+	for _, o := range allowedOrigins {
+		ao[o] = true
+	}
+	return &Interceptor{
+		AllowedOrigins: ao,
+		allowedHeaders: map[string]bool{},
+	}
+}
+
+// SetAllowedHeaders sets which headers should be allowed in the Access-Control-Allow-Headers
+// header. The headers are first canonicalized using textproto.CanonicalMIMEHeaderKey.
+func (it *Interceptor) SetAllowedHeaders(headers ...string) {
+	it.allowedHeaders = map[string]bool{}
+	for _, h := range headers {
+		if h == "*" {
+			continue
+		}
+		it.allowedHeaders[textproto.CanonicalMIMEHeaderKey(h)] = true
+	}
+}
+
+// Before handles the incoming request and sets headers or responds with an error
+// accordingly.
+func (it *Interceptor) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+	origin := r.Header.Get("Origin")
+	if origin != "" && !it.AllowedOrigins[origin] {
+		return w.ClientError(safehttp.StatusForbidden)
+	}
+	h := w.Header()
+	allowOrigin, err := h.Claim("Access-Control-Allow-Origin")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+	vary, err := h.Claim("Vary")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+
+	allowCredentials, err := h.Claim("Access-Control-Allow-Credentials")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+	if r.Header.Get("Cookie") != "" && !it.AllowCredentials {
+		return w.ClientError(safehttp.StatusForbidden)
+	}
+
+	var status safehttp.StatusCode
+	if r.Method() == safehttp.MethodOptions {
+		status = it.preflight(w, r)
+	} else {
+		status = it.request(w, r)
+	}
+
+	if status != 0 && status != safehttp.StatusNoContent {
+		if 400 <= status && status < 500 {
+			return w.ClientError(status)
+		} else {
+			return w.ServerError(status)
+		}
+	}
+
+	if origin != "" {
+		allowOrigin([]string{origin})
```

Comment:
We might need to revisit what the `Claim` returns. As it is right now, there is no way to *add* a header value, you always need to replace the whole thing.

Reply:
You could technically read the currently set header, add to it and then set it again. But I agree. this needs to be revised.